### PR TITLE
feat(Button): support for pre/post icons

### PR DESCRIPTION
**Pull request by atabel**
jira: WEB-625

### CODE REVIEW

File: src/button.tsx [hunk 1]
```diff
@@ -21,19 +21,20 @@ export const BUTTON_MIN_WIDTH = 136;
...
 
 const BORDER_PX = 1.5;
-const SPINNER_MARGIN_PX = 8;
+const ICON_MARGIN_PX = 8;
 const X_PADDING_PX = 16 - BORDER_PX;
 const Y_PADDING_PX = 12 - BORDER_PX;
 const X_SMALL_PADDING_PX = 12 - BORDER_PX;
 const Y_SMALL_PADDING_PX = 6 - BORDER_PX;
+const SMALL_ICON_SIZE = 20;
+const ICON_SIZE = 24;
```
Comment (yceballost) [inline src/button.tsx:31]: something happens in playroom snippet
https://ibit.ly/7ygh

**Review by yceballost — CHANGES_REQUESTED** [review]

File: src/button.tsx (hunk 1 above)
Comment (atabel) [reply src/button.tsx:31]: weird... it works fine in storybook. I'll take a look

File: src/button.tsx [hunk 2]
```diff
@@ -217,6 +219,62 @@ const useDangerButtonStyles = createUseStyles((theme) => ({
...
+        accText = [];
+    };
+    childrenArr.forEach((element, idx) => {
+        const isFirstChild = idx === 0;
+        const isLastChild = idx === length - 1;
+
+        const isIconElement =
+            React.isValidElement(element) &&
+            typeof element.type !== 'string' &&
+            element.type.name?.startsWith('Icon'); // all mistica icons start with Icon prefix
```
Comment (pladaria) [inline src/button.tsx:248]: You could be passing an element named IconWhatever that is not a mistica icon

File: src/button.tsx [hunk 3]
```diff
@@ -217,6 +219,62 @@ const useDangerButtonStyles = createUseStyles((theme) => ({
...
+                    key={resultChildrenArr.length}
+                    style={{
+                        display: 'flex',
+                        alignItems: 'center',
+                        marginLeft: isFirstChild ? 0 : ICON_MARGIN_PX,
+                        marginRight: isLastChild ? 0 : ICON_MARGIN_PX,
+                    }}
+                >
+                    {React.cloneElement(element, {size: pxToRem(sizeInPx)})}
+                </div>
```
Comment (pladaria) [inline src/button.tsx:266]: mistica icons accept a style prop. The div wrapper shouldn't be needed

File: src/button.tsx [hunk 4]
```diff
@@ -217,6 +219,62 @@ const useDangerButtonStyles = createUseStyles((theme) => ({
...
+    const childrenArr = React.Children.toArray(content);
+    const resultChildrenArr: Array<React.ReactNode> = [];
+    let accText: Array<React.ReactNode> = [];
+    const flushAccText = () => {
+        resultChildrenArr.push(
+            <React.Fragment key={resultChildrenArr.length}>{renderText(accText)}</React.Fragment>
+        );
+        accText = [];
+    };
+    childrenArr.forEach((element, idx) => {
```
Comment (pladaria) [inline src/button.tsx:241]: Instead of this, wouldn't be better to accept a startIcon / endIcon via props?

File: src/button.tsx [hunk 5]
```diff
@@ -21,19 +21,20 @@ export const BUTTON_MIN_WIDTH = 136;
 const transitionTiming = '0.3s cubic-bezier(0.77, 0, 0.175, 1)';
```
Comment (pladaria) [inline src/button.tsx:21]: are the button width changes in screenshots expected?

File: src/button.tsx (hunk 2 above)
Comment (atabel) [reply src/button.tsx:248]: that's fine, that element would likely be an icon too

File: src/button.tsx (hunk 4 above)
Comment (atabel) [reply src/button.tsx:241]: yes, but we already supported icons as children so I didn't want to change the component api

File: src/button.tsx (hunk 3 above)
Comment (atabel) [reply src/button.tsx:265]: As you said, it could be a non-mistica icon too, and the component may not have style prop. The extra div always works

**Review by aweell — APPROVED** [review]

File: src/button.tsx [hunk 6]
```diff
@@ -217,6 +221,59 @@ const useDangerButtonStyles = createUseStyles((theme) => ({
...
+
+        if (isIconElement) {
+            if (accText.length) {
+                flushAccText();
+            }
+            const sizeInPx = element.props.size ?? defaultIconSize;
+            resultChildrenArr.push(
+                <div
+                    key={resultChildrenArr.length}
+                    style={{
```
Comment (brtbrt) [inline src/button.tsx:257]: just for understanding purposes: why the style in place of a class in this case?

**Review by brtbrt — APPROVED** [review]

File: src/button.tsx [hunk 7]
```diff
@@ -21,19 +21,22 @@ export const BUTTON_MIN_WIDTH = 136;
...
 const BORDER_PX = 1.5;
-const SPINNER_MARGIN_PX = 8;
+const ICON_MARGIN_PX = 8;
 const X_PADDING_PX = 16 - BORDER_PX;
 const Y_PADDING_PX = 12 - BORDER_PX;
 const X_SMALL_PADDING_PX = 12 - BORDER_PX;
 const Y_SMALL_PADDING_PX = 6 - BORDER_PX;
+const ICON_SIZE = 24;
+const SMALL_ICON_SIZE = 20;
+const SPINNER_SIZE = 24;
```
Comment (yceballost) [inline src/button.tsx:31]:
```suggestion
const SPINNER_SIZE = 20;
```
Change this one to make the loading similar to icon size (icons have 2px padding inside, the real vector size is 20px)

Comment (atabel) [reply src/button.tsx:31]: hmmm... are you sure? it was 24 before my changes

File: src/button.tsx (hunk 6 above)
Comment (atabel) [reply src/button.tsx:257]: `marginLeft` and `marginRight` are dynamic styles, so it's easier and more convenient to use `style` here, although I could have 3 different classNames and combine them too

File: src/button.tsx (hunk 7 above)
Comment (yceballost) [reply src/button.tsx:31]: Yes, I know, we was talking @aweell and I to make this little change to keep a smooth consistency across icons and spinners

**Review by pladaria — APPROVED** [review]

**Comment by yceballost** [comment]
I think this ticket is solved with this pr
https://jira.tid.es/browse/WEB-598

**Pull request merged**